cpp/include/Ice/Proxy.h

```diff
@@ -264,11 +263,258 @@ namespace Ice
 class LocalException;
 class OutputStream;
 
+/**
+ * Helper template that supplies typed proxy factory functions.
+ * \headerfile Ice/Ice.h
+ */
+template<typename Prx, typename... Bases>
+class Proxy : public virtual Bases...
```

Comment:
Moved here because ObjectPrx now derives from this template class.

It uses the CRTP idiom: https://en.wikipedia.org/wiki/Curiously_recurring_template_pattern